### ISSUE TITLE
Allow metadata to be pushed instead of only pulled.

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#93]
In GitLab by @barry.nouwt.tno.nl on Jan 22, 2021, 09:56

(Instead of the currently implemented polling mechanism)

## What has to happen?
Several new POST and REACT knowledge interactions (meta) must be added for knowledge bases. It should post updates to interested parties (i.e., all other knowledge bases in the network) about this knowledge base and its knowledge interactions.

Because there can be different kinds of updates, we need to be able to distinguish:

- A new knowledge base is started
- An existing knowledge base is updated (e.g., new knowledge interaction)
- A knowledge base is removed

We want to communicate this distinction with the Communicative Act.

These three cases can each use the same graph pattern but the intended side effect of the messages is different.

The following subsections describe what happens in the three cases.

### New knowledge base
When a knowledge base's smart connector is first initialized, it should know about the other knowledge bases through the knowledge directory.
It should then send POST messages to those knowledge bases' meta knowledge interactions to introduce itself.

On the new knowledge base's side we do the POST:

POST requirements:

- None.

POST satisfactions:

- `NewKnowledgePurpose`

On all other knowledge bases' sides we do the REACT:

REACT requirements:

- `NewKnowledgePurpose` (or super class)

REACT satisfactions:

- None.

### Updated knowledge base
When a knowledge base is updated (e.g., a new knowledge interaction), it should send a POST message to inform its peers.

On the new knowledge base's side we do the POST:

POST requirements:

- None.

POST satisfactions:

- `ChangedKnowledgePurpose`

On all other knowledge bases' sides we do the REACT:

REACT requirements:

- `ChangedKnowledgePurpose` (or super class)

REACT satisfactions:

- None.

Questions:

- [x] Do we only need to include the changed resources, or the full bindings?
  - For now I will implement it with the full bindings. Later we could improve if we think it's necessary.

### Removed knowledge interaction
When a knowledge base is removed, it should tell its peers before leaving the network. For this, it sends a POST message.

On the to-be-terminated knowledge base's side we do the POST:

POST requirements:

- None.

POST satisfactions:

- `RemovedKnowledgePurpose`

On all other knowledge bases' sides we do the REACT:

REACT requirements:

- `RemovedKnowledgePurpose` (or super class)

REACT satisfactions:

- None.

Questions:

- [x] De we need to include the full bindings, or can we only give the URI of the knowledge interaction that is removed?
  - For now I'll just use the same graph pattern and fill all the bindings, even though I know most of them will be ignored.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Jan 29, 2021, 11:09

I think we first need to work on the CommunicativeAct for this. To distinguish between KnowledgeBases that became unavailable and KnowledgeBases that became available.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Feb 5, 2021, 08:54

marked this issue as related to #9

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Feb 23, 2021, 13:50

created merge request !66 to address this issue

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Feb 23, 2021, 14:22

marked the checklist item **Do we only need to include the changed resources, or the full bindings?** as completed

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Feb 23, 2021, 14:22

marked the checklist item **De we need to include the full bindings, or can we only give the URI of the knowledge interaction that is removed?** as completed

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Feb 25, 2021, 14:21

mentioned in commit 2f56c87dda291a756595b45ed59f65960f0d80db

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Feb 25, 2021, 14:30

Do we need a separate issue for supporting pushing of metadata to AdminUI, or did you already finish that in this issue?

**Comment by Sophietje** [comment]
In GitLab by @han.kruiger.tno.nl on Feb 25, 2021, 15:02

I (or we) did break the Admin UI, but that was because the graph pattern didn't match anymore. I updated the graph pattern to include the communicative act and it worked again :) https://gitlab.inesctec.pt/interconnect/knowledge-engine/-/merge_requests/67/diffs

But yes, it should also be possible in the future to implement the Admin UI with React Knowledge Interactions, but that will currently not work because of how we handle the meta messages.. Made an issue: #113

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Feb 25, 2021, 15:14

Very noble of you to claim that regression (is that the correct term? I think it is: https://en.wikipedia.org/wiki/Software_regression), but I was the one who added the Communicative Act and changed the Graph Pattern, but did not change the Admin UI graph pattern. I did, however, change the graph pattern of the `TestRequestMetadata`, so at least I did not do *everything* wrong.

This (again) shows the importance of our tests, because the failure of the tests pointed me towards fixing `TestRequestMetadata`.
